File: app/src/main/java/com/nfcalarmclock/activealarm/NacActiveAlarmActivity.java
package com.nfcalarmclock.activealarm;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Build;
import android.os.Bundle;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.Button;
import android.widget.RelativeLayout;
import android.widget.TextView;
import androidx.annotation.NonNull;
import com.nfcalarmclock.R;
import com.nfcalarmclock.alarm.NacAlarm;
import com.nfcalarmclock.nfc.NacNfc;
import com.nfcalarmclock.nfc.NacNfcTag;
import com.nfcalarmclock.shared.NacSharedConstants;
import com.nfcalarmclock.shared.NacSharedPreferences;
import com.nfcalarmclock.system.NacBundle;
import com.nfcalarmclock.system.NacIntent;
import com.nfcalarmclock.util.NacUtility;

/**
 * Activity to dismiss/snooze the alarm.
 */
@SuppressWarnings("RedundantSuppression")
public class NacActiveAlarmActivity
	extends Activity
	implements View.OnClickListener
{

	/**
	 * Receiver to stop the activity when the foreground service is done.
	 */
	private class StopActivityReceiver
		extends BroadcastReceiver
	{

		/**
		 */
		@Override
		public void onReceive(Context context, Intent intent)
		{
			String action = NacIntent.getAction(intent);
			if (action.equals(NacActiveAlarmActivity.ACTION_STOP_ACTIVITY))
			{
				finish();
			}
		}

	}

	/**
	 * Shared preferences.
	 */
	private NacSharedPreferences mSharedPreferences;

	/**
	 * Alarm.
	 */
	private NacAlarm mAlarm;

	/**
	 * Dismiss the activity action.
	 */
	public static final String ACTION_DISMISS_ACTIVITY =
		"com.nfcalarmclock.ACTION_DISMISS_ALARM_ACTIVITY";

	/**
	 * Stop the activity action.
	 */
	public static final String ACTION_STOP_ACTIVITY =
		"com.nfcalarmclock.ACTION_STOP_ALARM_ACTIVITY";

	/**
	 * Receiver to stop the activity.
	 */
    private StopActivityReceiver mStopReceiver;

	/**
	 * Cleanup NFC.
	 */
	private void cleanupNfc()
	{
		NacNfc.stop(this);
	}

	/**
	 * Cleanup the Stop receiver.
	 */
	private void cleanupStopReceiver()
	{
		StopActivityReceiver receiver = this.getStopReceiver();
		unregisterReceiver(receiver);
	}

	/**
	 * Dismiss the alarm.
	 */
	public void dismiss()
	{
		NacAlarm alarm = this.getAlarm();
		NacActiveAlarmService.dismissService(this, alarm);
	}

	/**
	 * Dismiss due to an NFC tag being scanned, and only if the NFC tag ID
	 * matches the saved alarm NFC tag ID. The exception to this is if the
	 * saved alarm NFC tag ID is empty.
	 * <p>
	 * The finish() method is not called because if the ACTION_DISMISS_ALARM
	 * intent is sent to the foreground service, then the foreground service will
	 * finish this activity.
	 */
	private void dismissFromNfcScan()
	{
		Intent intent = getIntent();
		NacAlarm alarm = this.getAlarm();
		NacNfcTag tag = new NacNfcTag(alarm, intent);

		if (tag.check(this))
		{
			NacActiveAlarmService.dismissServiceWithNfc(this, alarm);
		}
	}

	/**
	 * @return The alarm.
	 */
	private NacAlarm getAlarm()
	{
		return this.mAlarm;
	}

	/**
	 * @return The shared preferences.
	 */
	private NacSharedPreferences getSharedPreferences()
	{
		return this.mSharedPreferences;
	}

	/**
	 * @return The stop receiver.
	 */
	private StopActivityReceiver getStopReceiver()
	{
		return this.mStopReceiver;
	}

	/**
	 * Do not let the user back out of the activity.
	 */
	@Override
	public void onBackPressed()
	{
	}

	/**
	 */
	@Override
	public void onClick(View view)
	{
		NacSharedPreferences shared = this.getSharedPreferences();
		int id = view.getId();

		// Snooze
		if ((id == R.id.snooze) || ((id == R.id.act_alarm)
			&& shared.getEasySnooze()))
		{
			this.snooze();
		}
		// Dismiss
		else if (id == R.id.dismiss)
		{
			this.dismiss();
		}
	}

	/**
	 * Create the activity.
	 */
	@Override
	protected void onCreate(Bundle savedInstanceState)
	{
		super.onCreate(savedInstanceState);
		requestWindowFeature(Window.FEATURE_NO_TITLE);
		setContentView(R.layout.act_alarm);
		this.setAlarm(savedInstanceState);

		this.mSharedPreferences = new NacSharedPreferences(this);
		this.mStopReceiver = new StopActivityReceiver();

		// NFC tag was scanned
		if (this.wasNfcScanned())
		{
			this.dismissFromNfcScan();
		}

		// Check if alarm should be dismissed
		if (this.shouldDismissAlarm())
		{
			this.dismiss();
		}

		// Setup
		this.setupScreen();
		this.setupAlarmButtons();
		this.setupAlarmInfo();
	}

	/**
	 * NFC tag discovered so dismiss the dialog.
	 * <p>
	 * Note: Parent method must be called last. Causes issues with
	 * setSnoozeCount.
	 */
	@Override
	protected void onNewIntent(Intent intent)
	{
		setIntent(intent);
		this.dismissFromNfcScan();
	}

	/**
	 */
	@Override
	public void onPause()
	{
		super.onPause();

		this.cleanupNfc();
		this.cleanupStopReceiver();
	}

	/**
	 */
	@Override
	public void onResume()
	{
		super.onResume();

		this.setupNfc();
		this.setupAlarmInstructions();
		this.setupStopReceiver();
	}

	/**
	 * Save the alarm before the activity is killed.
	 */
	@Override
	public void onSaveInstanceState(@NonNull Bundle outState)
	{
		super.onSaveInstanceState(outState);

		NacAlarm alarm = this.getAlarm();

		// Save the alarm to the save instance state
		if (alarm != null)
		{
			outState.putParcelable(NacBundle.ALARM_PARCEL_NAME, alarm);
		}
	}

	/**
	 * Set the alarm.
	 */
	public void setAlarm(Bundle savedInstanceState)
	{
		NacAlarm alarm = NacIntent.getAlarm(getIntent());

		// Attempt to get the alarm from the saved instance state
		if (alarm == null)
		{
			alarm = NacBundle.getAlarm(savedInstanceState);
		}

		// Alarm is still null, finish the activity
		if (alarm == null)
		{
			finish();
		}

		// Set the alarm
		this.mAlarm = alarm;
	}

	/**
	 * Setup the snooze and dismiss buttons.
	 */
	public void setupAlarmButtons()
	{
		NacSharedPreferences shared = this.getSharedPreferences();
		NacAlarm alarm = this.getAlarm();
		RelativeLayout layout = findViewById(R.id.act_alarm);
		Button snoozeButton = findViewById(R.id.snooze);
		Button dismissButton = findViewById(R.id.dismiss);

		// NFC should be used so remove the dismiss button
		if (NacNfc.shouldUseNfc(this, alarm))
		{
			dismissButton.setVisibility(View.GONE);
		}
		// Show the dismiss button
		else
		{
			dismissButton.setVisibility(View.VISIBLE);
		}

		// Setup the buttons
		snoozeButton.setTextColor(shared.getThemeColor());
		dismissButton.setTextColor(shared.getThemeColor());
		layout.setOnClickListener(this);
		snoozeButton.setOnClickListener(this);
		dismissButton.setOnClickListener(this);
	}

	/**
	 * Setup the informational message at the bottom of the screen.
	 */
	public void setupAlarmInfo()
	{
		NacSharedPreferences shared = this.getSharedPreferences();
		NacAlarm alarm = this.getAlarm();

		// Alarm is present and the user wants to see alarm info
		if ((alarm != null) && shared.getShowAlarmInfo())
		{
			TextView name = findViewById(R.id.name);
			String alarmName = alarm.getNameNormalized();

			// Show alarm info
			name.setText(alarmName);
			name.setSelected(true);
		}
	}

	/**
	 * Setup the instruction message that appears just above the Snooze and
	 * Dismiss buttons.
	 */
	public void setupAlarmInstructions()
	{
		// Show NFC instructions
		if (!NacNfc.isEnabled(this))
		{
			TextView instructions = findViewById(R.id.instructions);
			instructions.setVisibility(View.GONE);
		}
	}

	/**
	 * Setup NFC.
	 */
	private void setupNfc()
	{
		NacAlarm alarm = this.getAlarm();

		// NFC is not enabled
		if (!NacNfc.isEnabled(this))
		{
			// NFC should be used, so prompt the user
			if (NacNfc.shouldUseNfc(this, alarm))
			{
				NacNfc.prompt(this);
			}
			// NFC does not need to be used. Return out of the method
			else
			{
				return;
			}
		}

		// NFC exists on the device. The device is NFC capable
		if (NacNfc.exists(this))
		{
			Intent intent = new Intent(this, NacActiveAlarmActivity.class)
				.addFlags(Intent.FLAG_RECEIVER_REPLACE_PENDING);

			// Start NFC
			NacNfc.start(this, intent);
		}
		// Unable to use NFC on the device
		else
		{
			NacSharedConstants cons = new NacSharedConstants(this);

			NacUtility.quickToast(this,
				cons.getErrorMessageNfcUnsupported());
		}
	}

	/**
	 * Setup the screen and handle the case when the device is locked.
	 */
	@SuppressWarnings("deprecation")
	public void setupScreen()
	{
		NacAlarm alarm = this.getAlarm();
		Window window = getWindow();
		boolean showWhenLocked = (alarm != null) && !alarm.shouldUseNfc();

		// Use updated method calls to control screen for APK >= 27
		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O_MR1)
		{
			setTurnScreenOn(true);
			setShowWhenLocked(showWhenLocked);
			//setShowWhenLocked(false);
		}
		else
		{
			window.addFlags(WindowManager.LayoutParams.FLAG_TURN_SCREEN_ON);

			// Add flag to show when locked
			if (showWhenLocked)
			{
				window.addFlags(WindowManager.LayoutParams.FLAG_SHOW_WHEN_LOCKED);
			}
			// Clear flag so app is not shown when locked
			else
			{
				window.clearFlags(WindowManager.LayoutParams.FLAG_SHOW_WHEN_LOCKED);
			}
		}

		// Keep screen on
		window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);

		// Allow lock screen when screen is turned on
		//window.addFlags(WindowManager.LayoutParams.FLAG_ALLOW_LOCK_WHILE_SCREEN_ON);
	}

	/**
	 * Setup the receiver for the Stop signal.
	 */
	@SuppressLint("UnspecifiedRegisterReceiverFlag")
	private void setupStopReceiver()
	{
		StopActivityReceiver receiver = this.getStopReceiver();
		IntentFilter filter = new IntentFilter(
			NacActiveAlarmActivity.ACTION_STOP_ACTIVITY);

		// Check if app needs to set the exported flag in order to indicate
		// that this app does not expect broadcasts from other apps on the
		// device
		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU)
		{
			int flags = Context.RECEIVER_NOT_EXPORTED;

			// Register to listen for the STOP broadcast for the activity
			registerReceiver(receiver, filter, flags);
		}
		else
		{
			// Register to listen for the STOP broadcast for the activity
			registerReceiver(receiver, filter);
		}
	}

	/**
	 * @return True if the alarm should be dismissed, and False otherwise.
	 */
	public boolean shouldDismissAlarm()
	{
		NacAlarm alarm = this.getAlarm();
		Intent intent = getIntent();
		String action = NacIntent.getAction(intent);

		// Check to see if the alarm should be dismissed and NFC does not need to be used
		return action.equals(NacActiveAlarmActivity.ACTION_DISMISS_ACTIVITY)
			&& !NacNfc.shouldUseNfc(this, alarm);
	}

	/**
	 * Snooze the alarm.
	 */
	public void snooze()
	{
		NacSharedPreferences shared = this.getSharedPreferences();
		NacAlarm alarm = this.getAlarm();

		// Unable to snooze. Show a toast indicating this to the user
		if (!alarm.canSnooze(shared))
		{
			NacSharedConstants cons = new NacSharedConstants(this);
			NacUtility.quickToast(this, cons.getErrorMessageSnooze());
			return;
		}

		// Snooze the alarm service
		NacActiveAlarmService.snoozeService(this, alarm);
	}

	/**
	 * @return True if an NFC tag was scanned, and False otherwise.
	 */
	public boolean wasNfcScanned()
	{
		Intent intent = getIntent();
		return NacNfc.wasScanned(intent);
	}

}

File: app/src/main/kotlin/com/nfcalarmclock/activealarm/NacActiveAlarmActivity.kt
package com.nfcalarmclock.activealarm

import android.animation.PropertyValuesHolder
import android.animation.ValueAnimator
import android.content.Context
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.view.MotionEvent
import android.view.View
import android.view.Window
import android.view.WindowManager
import androidx.activity.OnBackPressedCallback
import androidx.appcompat.app.AppCompatActivity
import com.nfcalarmclock.R
import com.nfcalarmclock.alarm.db.NacAlarm
import com.nfcalarmclock.nfc.NacNfc
import com.nfcalarmclock.shared.NacSharedPreferences
import com.nfcalarmclock.util.NacBundle
import com.nfcalarmclock.util.NacIntent
import com.nfcalarmclock.util.NacUtility.quickToast
import com.nfcalarmclock.util.enableActivityAlias
import dagger.hilt.android.AndroidEntryPoint
import kotlin.math.sqrt


/**
 * Activity to dismiss/snooze the alarm.
 */
@AndroidEntryPoint
class NacActiveAlarmActivity
	: AppCompatActivity()
{

	/**
	 * Shared preferences.
	 */
	private lateinit var sharedPreferences: NacSharedPreferences

	/**
	 * Alarm.
	 */
	private var alarm: NacAlarm? = null

	/**
	 * Original layout handler.
	 */
	private lateinit var originalLayoutHandler: NacOriginalLayoutHandler

	/**
	 * Swipe to dismiss layout handler.
	 */
	private lateinit var swipeLayoutHandler: NacSwipeLayoutHandler

	/**
	 * Listener for an alarm action, such as snooze or dismiss.
	 */
	val onAlarmActionListener: NacActiveAlarmLayoutHandler.OnAlarmActionListener =
		object: NacActiveAlarmLayoutHandler.OnAlarmActionListener
		{

			/**
			 * Called when the alarm should be snoozed.
			 */
			override fun onSnooze(alarm: NacAlarm)
			{
				// Snooze the alarm service. Whether the alarm is actually
				// snoozed is determined in the service
				NacActiveAlarmService.snoozeAlarmService(
					this@NacActiveAlarmActivity, alarm)
			}

			/**
			 * Called when the alarm should be dismissed.
			 */
			override fun onDismiss(alarm: NacAlarm)
			{
				// Dismiss the alarm service
				NacActiveAlarmService.dismissAlarmService(
					this@NacActiveAlarmActivity, alarm)
			}

		}

	/**
	 * Check if an alarm can be dismissed with NFC.
	 *
	 * @return True if an alarm can be dismissed with NFC, and False otherwise.
	 */
	private fun checkCanDismissWithNfc() : Boolean
	{
		// Get the NFC ID of the alarm and the intent
		val alarmNfcId = alarm!!.nfcTagId
		val intentNfcId = NacNfc.parseId(intent)

		// Compare the two NFC IDs. As long as nothing is null,
		// if the alarm NFC ID is empty, this is good, or
		// if the two NFC IDs are equal, this is also good
		return if ((alarm != null)
			&& (intentNfcId != null)
			&& (alarmNfcId.isEmpty() || (alarmNfcId == intentNfcId)))
		{
			true
		}
		// Something went wrong when comparing the NFC IDs
		else
		{
			// Show toast
			quickToast(this, R.string.error_message_nfc_mismatch)
			false
		}
	}

	/**
	 * Create the activity.
	 */
	override fun onCreate(savedInstanceState: Bundle?)
	{
		// Super
		super.onCreate(savedInstanceState)

		// Setup the window
		requestWindowFeature(Window.FEATURE_NO_TITLE)
		setContentView(R.layout.act_alarm_new)
		//setContentView(R.layout.act_alarm)

		// Set the alarm from the bundle
		setAlarm(savedInstanceState)

		// Set the member variables
		sharedPreferences = NacSharedPreferences(this)
		//originalLayoutHandler = NacOriginalLayoutHandler(this, alarm, onAlarmActionListener)
		swipeLayoutHandler = NacSwipeLayoutHandler(this, alarm, onAlarmActionListener)

		// Setup
		setupScreenOn()

		// Handle when the back button is pressed
		onBackPressedDispatcher.addCallback(this, object: OnBackPressedCallback(true)
		{
			override fun handleOnBackPressed()
			{
			}
		})
	}

	/**
	 * NFC tag discovered so dismiss the dialog.
	 *
	 * Note: Parent method must be called last. Causes issues with
	 * setSnoozeCount.
	 */
	override fun onNewIntent(intent: Intent)
	{
		// Super
		super.onNewIntent(intent)

		// Set the intent
		setIntent(intent)

		// Check if the alarm can be dismissed with NFC
		if (checkCanDismissWithNfc())
		{
			// Dismiss the alarm service with NFC
			NacActiveAlarmService.dismissAlarmServiceWithNfc(this, alarm!!)
		}
	}

	/**
	 * Called when the activity is paused.
	 */
	public override fun onPause()
	{
		// Super
		super.onPause()

		// Stop scanning for NFC
		NacNfc.stop(this)

		// Reenable the activity alias
		enableActivityAlias(this)
	}

	/**
	 * Called when the activity is resumed.
	 */
	public override fun onResume()
	{
		// Super
		super.onResume()

		// NFC tag was scanned. Check if can dismiss
		if (NacNfc.wasScanned(intent) && checkCanDismissWithNfc())
		{
			// Dismiss the alarm service with NFC
			NacActiveAlarmService.dismissAlarmServiceWithNfc(this, alarm!!)
		}

		// Setup
		setupNfc()
		//val x = findViewById<TextView>(R.id.title)
		//x.isSelected = true
	}

	/**
	 * Called when the window focus has changed. This is the best indicator of
	 * whether the activity is visible to the user, or not.
	 */
	override fun onWindowFocusChanged(hasFocus: Boolean)
	{
		// Super
		super.onWindowFocusChanged(hasFocus)

		println("WINDOW FOCUS CHANGE : $hasFocus")

		if (hasFocus)
		{
			// Start the swipe layout handler
			//originalLayoutHandler.start()
			swipeLayoutHandler.start()
		}
		else
		{
			// Stop the swipe layout handler
			swipeLayoutHandler.stop()
		}
	}

	/**
	 * Save the alarm before the activity is killed.
	 */
	public override fun onSaveInstanceState(outState: Bundle)
	{
		// Super
		super.onSaveInstanceState(outState)

		// Save the alarm to the save instance state
		if (alarm != null)
		{
			outState.putParcelable(NacBundle.ALARM_PARCEL_NAME, alarm)
		}
	}

	/**
	 * Set the alarm.
	 */
	private fun setAlarm(savedInstanceState: Bundle?)
	{
		// Attempt to get the alarm from the intent
		var intentAlarm = NacIntent.getAlarm(intent)

		// Unable to get the alarm from the intent
		if (intentAlarm == null)
		{
			// Attempt to get the alarm from the saved instance state
			intentAlarm = NacBundle.getAlarm(savedInstanceState)
		}

		// Alarm is still null, finish the activity
		if (intentAlarm == null)
		{
			finish()
		}

		// Check if the current alarm and the intent alarm are different
		// Set the alarm
		this.alarm = intentAlarm
	}

	/**
	 * Setup NFC.
	 */
	private fun setupNfc()
	{
		// NFC is not enabled
		if (!NacNfc.isEnabled(this))
		{
			// NFC should be used
			if (NacNfc.shouldUseNfc(this, alarm))
			{
				// Prompt the user
				NacNfc.prompt(this)
			}
			// NFC does not need to be used
			else
			{
				return
			}
		}

		// NFC exists on the device. The device is NFC capable
		if (NacNfc.exists(this))
		{
			// Create the intent
			val intent = Intent(this, NacActiveAlarmActivity::class.java)
				.addFlags(Intent.FLAG_RECEIVER_REPLACE_PENDING)

			// Start NFC
			NacNfc.start(this, intent)
		}
		// Unable to use NFC on this device
		else
		{
			// Show a toast
			quickToast(this, R.string.error_message_nfc_unsupported)
		}
	}

	/**
	 * Setup the screen and handle the case when the device is locked.
	 */
	@Suppress("deprecation")
	private fun setupScreenOn()
	{
		// Get whether the alarm should be shown when locked
		val showWhenLocked = (alarm != null) && !alarm!!.shouldUseNfc

		// Use updated method calls to control screen for APK >= 27
		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O_MR1)
		{
			setTurnScreenOn(true)
			setShowWhenLocked(showWhenLocked)
			//setShowWhenLocked(false);
		}
		else
		{
			// Turn on the screen
			window.addFlags(WindowManager.LayoutParams.FLAG_TURN_SCREEN_ON)

			// Add flag to show when locked
			if (showWhenLocked)
			{
				window.addFlags(WindowManager.LayoutParams.FLAG_SHOW_WHEN_LOCKED)
			}
			// Do not show when locked
			else
			{
				window.clearFlags(WindowManager.LayoutParams.FLAG_SHOW_WHEN_LOCKED)
			}
		}

		// Keep screen on
		window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)

		// Allow lock screen when screen is turned on
		//window.addFlags(WindowManager.LayoutParams.FLAG_ALLOW_LOCK_WHILE_SCREEN_ON);
	}

	companion object
	{

		/**
		 * Stop the activity action.
		 */
		private const val ACTION_STOP_ACTIVITY = "com.nfcalarmclock.ACTION_STOP_ALARM_ACTIVITY"

		/**
		 * Create an intent that will be used to start the Alarm activity.
		 *
		 * @param context A context.
		 * @param alarm   An alarm.
		 *
		 * @return The Alarm activity intent.
		 */
		fun getStartIntent(context: Context, alarm: NacAlarm?): Intent
		{
			// Create the intent and its flags
			val intent = Intent(context, NacActiveAlarmActivity::class.java)
			val flags = (Intent.FLAG_ACTIVITY_NEW_TASK
				or Intent.FLAG_ACTIVITY_CLEAR_TASK)

			// Add the flags to the intent
			intent.addFlags(flags)

			return NacIntent.addAlarm(intent, alarm)
		}

		/**
		 * Start the alarm activity with the given alarm.
		 */
		fun startAlarmActivity(context: Context, alarm: NacAlarm?)
		{
			// Create the intent
			val intent = getStartIntent(context, alarm)

			// Start the activity
			context.startActivity(intent)
		}

		/**
		 * Stop the alarm activity
		 */
		fun stopAlarmActivity(context: Context)
		{
			// Create the intent and its flags
			val intent = Intent(context, NacActiveAlarmActivity::class.java)
			val flags = (Intent.FLAG_ACTIVITY_NEW_TASK
				or Intent.FLAG_ACTIVITY_CLEAR_TASK)

			// Add the flags to the intent
			intent.addFlags(flags)
			intent.action = ACTION_STOP_ACTIVITY

			// Start the activity
			context.startActivity(intent)
		}

	}

}

/**
 * Calculate the radius of where a user is touching with respect to a view.
 */
fun View.calculateTouchRadius(motionEvent: MotionEvent): Int
{
	// Determine the X and Y from the reference of the origin, as
	// opposed to the top-left corner
	val xOrigin = this.layoutParams.width / 2
	val yOrigin = this.layoutParams.height / 2
	val newX = motionEvent.x - xOrigin
	val newY = motionEvent.y - yOrigin

	// Calculate the radius of where the user's finger is wrt the view
	return sqrt(newX*newX + newY*newY).toInt()
}

/**
 * Set the size of a view.
 */
fun View.setViewSize(size: Int)
{
	// Set the width and height
	this.layoutParams.width = size
	this.layoutParams.height = size

	// Force a layout refresh
	this.requestLayout()
}

/**
 * Set the size of a view from an animator.
 */
fun View.setViewSizeFromAnimator(valueAnimator: ValueAnimator)
{
	// Get the value from the animator
	val value = valueAnimator.animatedValue as Int

	// Set the size of the view
	this.setViewSize(value)
}

/**
 * Extension function to convert the values of a property holder to an Int.
 */
fun PropertyValuesHolder.getIntValues(): List<Int>
{
	return this.toString()
		.replace(":", "")
		.trim()
		.replace("\\s+".toRegex(), " ")
		.split("\\s".toRegex())
		.map { it.toInt() }
}
